cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -0,0 +1,103 @@
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteTaskExecutor.class);
+  private final RemoteClient remoteClient;
+  private final RetryStrategy retryStrategy;
+
+  public RemoteTaskExecutor(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.TASK_WORKER,
+                                         new DefaultHttpRequestConfig(false),
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+  }
+
+  /**
+   * Sends the {@link RunnableTaskRequest} to a remote worker and returns the result.
+   * Retries sending the request if the workers are busy.
+   *
+   * @param runnableTaskRequest {@link RunnableTaskRequest} with details of task
+   * @return byte[] response from remote task
+   * @throws Exception returned by remote task if any
+   */
+  public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws Exception {
+    HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.POST, TASK_WORKER_URL);
+    HttpRequest httpRequest = requestBuilder.withBody(GSON.toJson(runnableTaskRequest)).build();
+    return Retries.callWithRetries(() -> {
+      try {
+        HttpResponse httpResponse = remoteClient.execute(httpRequest);
+        if (httpResponse.getResponseCode() == HttpResponseStatus.TOO_MANY_REQUESTS.code()) {
+          throw new RetryableException(
+            String.format("Received response code %s for %s", httpResponse.getResponseCode(),
+                          runnableTaskRequest.getClassName()));
+        }
+        if (httpResponse.getResponseCode() != 200) {
+          BasicThrowable basicThrowable = GSON
+            .fromJson(new String(httpResponse.getResponseBody()), BasicThrowable.class);
+          throw createExceptionObject(basicThrowable);
+        }
+        return httpResponse.getResponseBody();
+      } catch (NoRouteToHostException | ServiceUnavailableException e) {
+        throw new RetryableException(
+          String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
+      }
+    }, retryStrategy);
+  }
+
+  private Exception createExceptionObject(@Nullable BasicThrowable basicThrowable) {
+    if (basicThrowable == null) {
+      return null;
+    }
+    String exceptionMessage = String.format("Throwable %s thrown with message %s",
+                                            basicThrowable.getClassName(), basicThrowable.getMessage());
+    RemoteTaskException remoteTaskException = new RemoteTaskException(exceptionMessage,
+                                                                      createExceptionObject(basicThrowable.getCause()));
+    remoteTaskException.setStackTrace(basicThrowable.getStackTraces());
+    return new Exception("Exception from remote task", remoteTaskException);
```

Comment:
Wrapping it here in Exception so that we can trace the exception to this method. RemoteTaskException's stacktrace is being set to the stacktrace received from remote task. See Terence's comment  - https://github.com/cdapio/cdap/pull/13383#discussion_r633285374